### PR TITLE
Add support for extended ASCII in password generator

### LINKED ISSUE #343: Add support for high ANSI characters in the password generator
This is something I miss from when I used to use KeePass on my windows partition. The current character set is great, but the addition of high ANSI characters would increase the strength of a password and would make KeePassXC even more awesome.

### PR DESCRIPTION
**Pull request by TheZ3ro**
Resolve #343 

## Description

This adds support for extended ASCII character in the password generator, like KeePass.

## How has this been tested?

Manually and with Travis

## Screenshots (if appropriate):
istantanea_2017-04-28_22-05-01

## Types of changes

- ✅ New feature (non-breaking change which adds functionality)

## Checklist:

- ✅ I have read the **CONTRIBUTING** document. **[REQUIRED]**
- ✅ My code follows the code style of this project. **[REQUIRED]**
- ✅ All new and existing tests passed. **[REQUIRED]**
- ✅ I have compiled and verified my code with `-DWITH_ASAN=ON`. **[REQUIRED]**

### CODE REVIEW

File: src/gui/PasswordGeneratorWidget.cpp [hunk 1]
```diff
@@ -92,6 +92,7 @@ void PasswordGeneratorWidget::loadSettings()
     m_ui->checkBoxUpper->setChecked(config()->get("generator/UpperCase", true).toBool());
     m_ui->checkBoxNumbers->setChecked(config()->get("generator/Numbers", true).toBool());
     m_ui->checkBoxSpecialChars->setChecked(config()->get("generator/SpecialChars", false).toBool());
+    m_ui->checkBoxExtASCII->setChecked(config()->get("generator/EASCII", false).toBool());
```
Comment (louib) [inline src/gui/PasswordGeneratorWidget.cpp]: @TheZ3ro I think we could call this one `checkBoxExtendedASCII` or `checkBoxEASCII`

**Comment by louib** [comment]
@TheZ3ro I guess we could also blacklist those, yeah.
Is there any unit tests to add? I don't see a test file for password generators.

**Comment by TheZ3ro** [comment]
@louib I was also wondering about what to do. I think we can add a new test file for them

File: src/gui/PasswordGeneratorWidget.cpp (hunk 1 above)
Comment (TheZ3ro) [inline src/gui/PasswordGeneratorWidget.cpp]: @louib isn't `Ext` an abbreviation for `Extended`? I will change this to `checkBoxExtendedASCII` if `Ext` don't fit.

Comment (louib) [inline src/gui/PasswordGeneratorWidget.cpp]: @TheZ3ro I don't mind the extra characters to make it explicit, but maybe that's just me.

**Comment by louib** [comment]
@TheZ3ro also, I searched for look-alike inside of the EASCII group, but I did not pay attention to possible look-alike across groups with EASCII enabled, which the other classes seem to be handling. So we should see if there's any more look-alikes to exclude.

**Comment by TheZ3ro** [comment]
@louib this is the code for EASCII and SpecialChar from KeePass source:

```
if(m_strHighAnsi == null)
{
	StringBuilder sbHighAnsi = new StringBuilder();
	// [U+0080, U+009F] are C1 control characters,
	// U+00A0 is non-breaking space
	for(char ch = '\u00A1'; ch <= '\u00AC'; ++ch)
		sbHighAnsi.Append(ch);
	// U+00AD is soft hyphen (format character)
	for(char ch = '\u00AE'; ch < '\u00FF'; ++ch)
		sbHighAnsi.Append(ch);
	sbHighAnsi.Append('\u00FF');

	m_strHighAnsi = sbHighAnsi.ToString();
}

if(m_strSpecial == null)
{
	PwCharSet pcs = new PwCharSet(false);
	pcs.AddRange('!', '/');
	pcs.AddRange(':', '@');
	pcs.AddRange('[', '`');
	pcs.Add(@"|~");
	pcs.Remove(@"-_ ");
	pcs.Remove(PwCharSet.Brackets);

	m_strSpecial = pcs.ToString();
}
```

This is the only look alike
```
public const string LookAlike = @"O0l1I|";
```

File: src/core/PasswordGenerator.h [hunk 2]
```diff
@@ -32,7 +32,8 @@ class PasswordGenerator
         LowerLetters      = 0x1,
         UpperLetters      = 0x2,
         Numbers           = 0x4,
-        SpecialCharacters = 0x8
+        SpecialCharacters = 0x8,
+        EASCII            = 0x16
```
Comment (louib) [inline src/core/PasswordGenerator.h]: @TheZ3ro oops :open_mouth:

File: src/core/PasswordGenerator.h [hunk 3]
```diff
@@ -33,7 +33,7 @@ class PasswordGenerator
         UpperLetters      = 0x2,
         Numbers           = 0x4,
         SpecialCharacters = 0x8,
-        EASCII            = 0x16
+        EASCII            = 0x10
```
Comment (louib) [inline src/core/PasswordGenerator.h]: @TheZ3ro :open_mouth: my bad, nice catch!

File: src/gui/PasswordGeneratorWidget.cpp (hunk 1 above)
Comment (droidmonkey) [inline src/gui/PasswordGeneratorWidget.cpp]: I vote for the name as is. It's descriptive to me.

**Comment by TheZ3ro** [comment]
@louib how do you think we should test the password generator? Generating random password and checking if they respect the char restriction from their groups?

Generating one password doesn't really tell you if the generator is working properly (maybe the output password is valid only because it didn't selected an invalid char for this specific generation)
Should I generate N password?

And also, the restriction checking mechanism is based on the generator itself so it will pass the tests also in case where it should fail.

Do we need tests for the generator itself?

**Comment by louib** [comment]
> Generating random password and checking if they respect the char restriction from their groups?

Pretty much what I was thinking, but I guess we can go ahead without testing this.

File: src/gui/PasswordGeneratorWidget.cpp (hunk 1 above)
Comment (eternaltyro) [inline src/gui/PasswordGeneratorWidget.cpp]: I agree with @louib  too. Less confusion that way.

**Pull request merged**